contracts/guilds/GuildManager.cairo

```diff
@@ -0,0 +1,168 @@
+%lang starknet
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import deploy, get_caller_address
+from contracts.lib.math_utils import array_product
+
+from starkware.cairo.common.bool import TRUE, FALSE
+from contracts.lib.role_new import Role
+
+from contracts.interfaces.IGuildCertificate import IGuildCertificate
+
+#
+# Constants
+#
+
+const INITIALIZE_SELECTOR = 215307247182100370520050591091822763712463273430149262739280891880522753123
+
+#
+# Structs
+#
+
+struct ProxyDeployData:
+    member implementation : felt
+    member selector : felt
+    member calldata_len : felt
+    member calldata : felt*
+end
+
+#
+# Storage variables
+#
+
+@storage_var
+func salt() -> (value : felt):
+end
+
+@storage_var
+func guild_proxy_class_hash() -> (value : felt):
+end
+
+@storage_var
+func guild_class_hash() -> (value : felt):
+end
+
+@storage_var
+func guild_contract_count() -> (res : felt):
+end
+
+@storage_var
+func guild_contracts(index : felt) -> (value : felt):
+end
+
+#
+# Events
+#
+
+@event
+func guild_contract_deployed(name : felt, master : felt, contract_address : felt):
+end
+
+@constructor
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    guild_proxy_class_hash_ : felt, guild_class_hash_ : felt
+):
+    guild_proxy_class_hash.write(value=guild_proxy_class_hash_)
+    guild_class_hash.write(value=guild_class_hash_)
+    return ()
+end
+
+@external
+func deploy_guild_proxy_contract{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    name : felt, guild_certificate : felt
+):
+    let (current_salt) = salt.read()
+    let (proxy_class_hash) = guild_proxy_class_hash.read()
+    let (class_hash) = guild_class_hash.read()
+    let (caller_address) = get_caller_address()
+    let (calldata : felt*) = alloc()
+    assert calldata[0] = class_hash
+    assert calldata[1] = INITIALIZE_SELECTOR
+    assert calldata[2] = 3
+    assert calldata[3] = name
+    assert calldata[4] = caller_address
+    assert calldata[5] = guild_certificate
+    let (contract_address) = deploy(
+        class_hash=proxy_class_hash,
+        contract_address_salt=current_salt,
+        constructor_calldata_size=6,
+        constructor_calldata=calldata,
+    )
+    salt.write(value=current_salt + 1)
+
+    guild_contract_deployed.emit(
+        name=name, master=caller_address, contract_address=contract_address
+    )
+
+    let (contract_count) = guild_contract_count.read()
+    guild_contracts.write(contract_count, contract_address)
+    guild_contract_count.write(contract_count + 1)
+
+    IGuildCertificate.mint(
+        contract_address=guild_certificate,
+        to=caller_address,
+        guild=contract_address,
+        role=Role.ADMIN,
+    )
+
+    return ()
+end
+
+@external
+func check_valid_contract{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```

Comment:
Whenever a certificate is minted/burned it needs to make sure the call came from a guild. The `guild_manager` stores a list of the guilds that it deploys (as a factory), and `check_valid_contract` will check that an address exists in this guild list. The `guild_certificate` calls this passing the caller address and if `false` is returned then `guild_certificate` will revert.

Reply:
Ok. I think I got it, but this is quite convoluted in many ways 😕

First of, if the call is *correct* (i.e. it came from a guild), the return value `check_product` is 0, which is also the value for `FALSE` - so that's one weird thing (FALSE signifying correct state). 

Next, it could actually be made more efficient if you'd return early from `_check_valid_contract` in the case `check` is actually 0 there, because the the whole product will be zero.

There's also the very theoretical chance of `check_guild * check_manager` to be 0 because of finite field math...but yeah, that's very unlikely.

However, overall, I don't think this is the optimal solution - as I mentioned in the other comment, this is also an `O(n)` operation which is not ideal in a game that can run for a long time and many many guilds can be created. IMO it would be better if you'd store the address directly. Then the lookup will become a constant time operation. 

Does that make sense? Maybe I'm also missing something. Curious to read your thoughts.

Reply:
Makes sense and your points are totally valid. If I understand correctly, it would be better just to have an `is_guild` storage variable with address mapped to `TRUE` or `FALSE`. That way one address and response will be checked every time rather than going through a potentially infinite list.

Pretty clear cut, will make the change. Other methods in the contracts used a similar principle like checking what tokens you own, I will also change these to be in line. Many thanks 🙌